src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,858 @@
+#include "cardinality_estimator.hpp"
+
+#include <iostream>
+#include <memory>
+
+#include "column_statistics.hpp"
+#include "expression/abstract_expression.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/lqp_subquery_expression.hpp"
+#include "expression/value_expression.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/alias_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/limit_node.hpp"
+#include "logical_query_plan/mock_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "logical_query_plan/sort_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "logical_query_plan/validate_node.hpp"
+#include "operators/operator_join_predicate.hpp"
+#include "operators/operator_scan_predicate.hpp"
+#include "resolve_type.hpp"
+#include "statistics/cardinality_estimation_cache.hpp"
+#include "statistics/column_statistics.hpp"
+#include "statistics/statistics_objects/equal_distinct_count_histogram.hpp"
+#include "statistics/statistics_objects/generic_histogram.hpp"
+#include "statistics/statistics_objects/generic_histogram_builder.hpp"
+#include "statistics/statistics_objects/single_bin_histogram.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "table_statistics.hpp"
+#include "utils/assert.hpp"
+
+namespace {
+
+using namespace opossum;  // NOLINT
+
+template <typename T>
+std::optional<float> estimate_null_ratio_of_column(const TableStatistics& table_statistics,
+                                                   const ColumnStatistics<T>& column_statistics) {
+  // If the column has an explicit null value ratio associated with it, we can just use that
+  if (column_statistics.null_value_ratio) {
+    return column_statistics.null_value_ratio->ratio;
+  }
+
+  // Otherwise derive the NVR from the total count of an histogram (which excludes NULLs) and the TableStatistics
```

Comment:
Do we need this abbreviation? Especially because the function does not use the "V"?
Also: a**n** histogram